gedi/process_l2a.py

```diff
@@ -0,0 +1,290 @@
+    df : pd.DataFrame
+        The DataFrame, assumed to be the output of gedi_L2A_to_df, for which
+        canopy metrics should be calculated. The provided df must contain a column
+        labeled "rh" which is array-valued.
+    canopy_threshold : float
+        The minimum value for a return to be considered a "canopy" return.
+
+    Side Effects
+    ------------
+    Modifies the provided pd.DataFrame in-place to add canopy metrics.
+
+    Returns
+    -------
+    None
+    """
+    rh = np.stack(df.rh.to_numpy())
+    canopy_returns = np.ma.masked_less(rh, canopy_threshold)
+
+    canopy_max = np.max(canopy_returns, axis=1)
+    canopy_min = np.min(canopy_returns, axis=1)
+    canopy_std = np.std(canopy_returns, axis=1)
+    canopy_avg = np.average(canopy_returns, axis=1)
+    dns = canopy_returns.count(axis=1)  # % all returns >= canopy threshold
+
+    canopy_returns = np.ma.filled(canopy_returns, np.nan)
+    canopy_p10 = _compute_nan_percentile(canopy_returns, 10)
+    canopy_p25 = _compute_nan_percentile(canopy_returns, 25)
+    canopy_p50 = _compute_nan_percentile(canopy_returns, 50)
+    canopy_p75 = _compute_nan_percentile(canopy_returns, 75)
+    canopy_p90 = _compute_nan_percentile(canopy_returns, 90)
+
+    d01 = np.ma.masked_outside(rh, canopy_threshold, 5).count(
+        axis=1
+    )  # % returns >= canopy, <=5m
+    d02 = np.ma.masked_outside(rh, 5, 10).count(axis=1)  # % returns >= 5m, <=10m
+    d03 = np.ma.masked_outside(rh, 10, 20).count(axis=1)  # % returns >= 10m, <=20m
+    d04 = np.ma.masked_outside(rh, 20, 30).count(axis=1)  # % returns >= 20m, <=30m
+
+    df["canopy_max"] = pd.Series(canopy_max)
+    df["canopy_min"] = pd.Series(canopy_min)
+    df["canopy_std"] = pd.Series(canopy_std)
+    df["canopy_avg"] = pd.Series(canopy_avg)
+    df["canopy_p10"] = pd.Series(canopy_p10)
+    df["canopy_p25"] = pd.Series(canopy_p25)
+    df["canopy_p50"] = pd.Series(canopy_p50)
+    df["canopy_p75"] = pd.Series(canopy_p75)
+    df["canopy_p90"] = pd.Series(canopy_p90)
+    df["dns"] = pd.Series(dns)
+    df["d01"] = pd.Series(d01)
+    df["d02"] = pd.Series(d02)
+    df["d03"] = pd.Series(d03)
+    df["d04"] = pd.Series(d04)
+
+
+def df_to_geojson(df, outfile):
+    """
+    Convert pandas dataframe to GeoJSON file and save as given file name and path.
+    This file type can be useful when working locally in GIS software.
+
+    Parameters
+    ----------
+    df : pd.DataFrame
+        The DataFrame, assumed to be the output of gedi_L2A_to_df, with or without
+        canopy height metrics appended.
+    outfile : str
+        The location and filename of output json file.
+
+    Returns
+    -------
+    None
+    """
+    df["geometry"] = df.apply(
+        lambda row: Point(row.lon_lowestmode, row.lat_lowestmode), axis=1
+    )
+    GeoDF = gp.GeoDataFrame(df)
+    GeoDF = GeoDF.drop(columns=["lat_lowestmode", "lon_lowestmode"])
+    GeoDF.to_file(outfile, driver="GeoJSON")
+
+
+if __name__ == "__main__":
```

Comment:
When I started writing this, there wasn't much going on and a separate `main` function felt unnecessary, but perhaps it's time to revisit that assumption.